mediatek: add linux 6.1 testing kernel

dangowrt opened the pull request:
With Linux 6.1 many of our downstream patches and out-of-tree files can be removed or at least replaced by backported upstream commits.
 * Adapt old RTK switch driver to build with Linux 6.1
 * Add updated configuration for all subtargets and set KERNEL_TESTING_PATCHVER:=6.1

Compile tested only!

dangowrt committed:
mediatek: expose testing Linux 6.1 testing kernel

Set KERNEL_TESTING_PATCHVER:=6.1 to allow building all subtargets with
Linux 6.1.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
mediatek: prepare old rtk switch driver for use with Linux 6.1

The old RealTek RTL8367S switch driver which is used for some MT7622
devices needs to be modified to no longer free the GPIO after reset
has completed.
This is due to Linux 5.19 removing devm_gpio_free via commit
2b038e786f83 ("gpiolib: devres: Get rid of unused devm_gpio_free()")

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

Ansuel commented:
@dangowrt nitpick free free in the first commit description

robimarko commented:
It would be probably best to convert the driver to gpiod instead and stop using the legacy API.

rsalvaterra commented:
Oh, goody! You beat me to it, as I expected, @dangowrt. Will give it a spin ASAP, I've been looking forward to this. :wink:

Ansuel commented:
missing kconfig.

rsalvaterra commented:
Lost the compiler output, but I got incompatible pointer type errors building the MT8183 clock driver. I don't need it for MT7622, but still… :confused:

bmork commented:
Successfully run-tested this on the filogic based ZyXEL EX5700 ( https://github.com/openwrt/openwrt/pull/12276 - would be nice to finally have that merged too. hint, hint :)

bmork commented:
> [    3.477730] ubi0 error: ubi_read_volume_table: not enough PEBs, required 1974, available 992

Looks like you've got only half the expected number of PEBs for some reason.  

But looking at the AX4200 DTS, it seems the UBI_DEV partition is only ~250MB.  This makes the 992 number reasonable.  So how did you end up with volumes claiming 1974?

What does the log normally say on a successful boot?

Gingernut1978 commented:
> > [    3.477730] ubi0 error: ubi_read_volume_table: not enough PEBs, required 1974, available 992
> 
> Looks like you've got only half the expected number of PEBs for some reason.
> 
> But looking at the AX4200 DTS, it seems the UBI_DEV partition is only ~250MB. This makes the 992 number reasonable. So how did you end up with volumes claiming 1974?
> 
> What does the log normally say on a successful boot?

Here's a bootlog on 5.15

``[    0.000000] Booting Linux on physical CPU 0x0000000000 [0x410fd034]
[    0.000000] Linux version 5.15.102 (asim@ASIM-MS-7D20) (aarch64-openwrt-linux-musl-gcc (OpenWrt GCC 12.2.0 r21874+2-f183ce35b8) 12.2.0, GNU ld (GNU Binutils) 2.40.0) #0 SMP Fri Mar 24 19:30:29 2023
[    0.000000] Machine model: ASUS TUF-AX4200
[    0.000000] Zone ranges:
[    0.000000]   DMA      [mem 0x0000000040000000-0x000000005fffffff]
[    0.000000]   DMA32    empty
[    0.000000]   Normal   empty
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000040000000-0x0000000042ffffff]
[    0.000000]   node   0: [mem 0x0000000043000000-0x000000004302ffff]
[    0.000000]   node   0: [mem 0x0000000043030000-0x000000004fbfffff]
[    0.000000]   node   0: [mem 0x000000004fc00000-0x000000004ffbffff]
[    0.000000]   node   0: [mem 0x000000004ffc0000-0x000000005fffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000040000000-0x000000005fffffff]
[    0.000000] psci: probing for conduit method from DT.
[    0.000000] psci: PSCIv1.1 detected in firmware.
[    0.000000] psci: Using standard PSCI v0.2 function IDs
[    0.000000] psci: MIGRATE_INFO_TYPE not supported.
[    0.000000] psci: SMC Calling Convention v1.2
[    0.000000] percpu: Embedded 17 pages/cpu s29592 r8192 d31848 u69632
[    0.000000] pcpu-alloc: s29592 r8192 d31848 u69632 alloc=17*4096
[    0.000000] pcpu-alloc: [0] 0 [0] 1 [0] 2 [0] 3
[    0.000000] Detected VIPT I-cache on CPU0
[    0.000000] CPU features: detected: GIC system register CPU interface
[    0.000000] CPU features: kernel page table isolation disabled by kernel configuration
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 129024
[    0.000000] Kernel command line: ubi.mtd=UBI_DEV
[    0.000000] Dentry cache hash table entries: 65536 (order: 7, 524288 bytes, linear)
[    0.000000] Inode-cache hash table entries: 32768 (order: 6, 262144 bytes, linear)
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 497888K/524288K available (8064K kernel code, 882K rwdata, 2124K rodata, 384K init, 286K bss, 26400K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=64, Order=0-3, MinObjects=0, CPUs=4, Nodes=1
[    0.000000] rcu: Hierarchical RCU implementation.
[    0.000000]  Tracing variant of Tasks RCU enabled.
[    0.000000] rcu: RCU calculated value of scheduler-enlistment delay is 10 jiffies.
[    0.000000] NR_IRQS: 64, nr_irqs: 64, preallocated irqs: 0
[    0.000000] GICv3: GIC: Using split EOI/Deactivate mode
[    0.000000] GICv3: 640 SPIs implemented
[    0.000000] GICv3: 0 Extended SPIs implemented
[    0.000000] GICv3: Distributor has no Range Selector support
[    0.000000] Root IRQ handler: gic_handle_irq
[    0.000000] GICv3: 16 PPIs implemented
[    0.000000] GICv3: CPU0: found redistributor 0 region 0:0x000000000c080000
[    0.000000] arch_timer: cp15 timer(s) running at 13.00MHz (phys).
[    0.000000] clocksource: arch_sys_counter: mask: 0xffffffffffffff max_cycles: 0x2ff89eacb, max_idle_ns: 440795202429 ns
[    0.000000] sched_clock: 56 bits at 13MHz, resolution 76ns, wraps every 4398046511101ns
[    0.000125] Calibrating delay loop (skipped), value calculated using timer frequency.. 26.00 BogoMIPS (lpj=130000)
[    0.000133] pid_max: default: 32768 minimum: 301
[    0.000207] Mount-cache hash table entries: 1024 (order: 1, 8192 bytes, linear)
[    0.000215] Mountpoint-cache hash table entries: 1024 (order: 1, 8192 bytes, linear)
[    0.001097] rcu: Hierarchical SRCU implementation.
[    0.001178] dyndbg: Ignore empty _ddebug table in a CONFIG_DYNAMIC_DEBUG_CORE build
[    0.001367] smp: Bringing up secondary CPUs ...
[    0.001607] Detected VIPT I-cache on CPU1
[    0.001629] GICv3: CPU1: found redistributor 1 region 0:0x000000000c0a0000
[    0.001654] CPU1: Booted secondary processor 0x0000000001 [0x410fd034]
[    0.001916] Detected VIPT I-cache on CPU2
[    0.001928] GICv3: CPU2: found redistributor 2 region 0:0x000000000c0c0000
[    0.001938] CPU2: Booted secondary processor 0x0000000002 [0x410fd034]
[    0.002166] Detected VIPT I-cache on CPU3
[    0.002176] GICv3: CPU3: found redistributor 3 region 0:0x000000000c0e0000
[    0.002185] CPU3: Booted secondary processor 0x0000000003 [0x410fd034]
[    0.002209] smp: Brought up 1 node, 4 CPUs
[    0.002222] SMP: Total of 4 processors activated.
[    0.002225] CPU features: detected: 32-bit EL0 Support
[    0.002229] CPU features: detected: CRC32 instructions
[    0.002300] CPU: All CPU(s) started at EL2
[    0.002311] alternatives: patching kernel code
[    0.004270] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.004290] futex hash table entries: 1024 (order: 4, 65536 bytes, linear)
[    0.004404] pinctrl core: initialized pinctrl subsystem
[    0.004842] NET: Registered PF_NETLINK/PF_ROUTE protocol family
[    0.005048] DMA: preallocated 128 KiB GFP_KERNEL pool for atomic allocations
[    0.005069] DMA: preallocated 128 KiB GFP_KERNEL|GFP_DMA pool for atomic allocations
[    0.005086] DMA: preallocated 128 KiB GFP_KERNEL|GFP_DMA32 pool for atomic allocations
[    0.005305] thermal_sys: Registered thermal governor 'fair_share'
[    0.005308] thermal_sys: Registered thermal governor 'bang_bang'
[    0.005311] thermal_sys: Registered thermal governor 'step_wise'
[    0.005314] thermal_sys: Registered thermal governor 'user_space'
[    0.005474] ASID allocator initialised with 65536 entries
[    0.005709] pstore: Registered ramoops as persistent store backend
[    0.005713] ramoops: using 0x10000@0x42ff0000, ecc: 0
[    0.012973] cryptd: max_cpu_qlen set to 1000
[    0.014616] SCSI subsystem initialized
[    0.014693] libata version 3.00 loaded.
[    0.015410] clocksource: Switched to clocksource arch_sys_counter
[    0.015802] NET: Registered PF_INET protocol family
[    0.015879] IP idents hash table entries: 8192 (order: 4, 65536 bytes, linear)
[    0.016150] tcp_listen_portaddr_hash hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.016162] Table-perturb hash table entries: 65536 (order: 6, 262144 bytes, linear)
[    0.016169] TCP established hash table entries: 4096 (order: 3, 32768 bytes, linear)
[    0.016190] TCP bind hash table entries: 4096 (order: 4, 65536 bytes, linear)
[    0.016233] TCP: Hash tables configured (established 4096 bind 4096)
[    0.016289] UDP hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.016301] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.016374] NET: Registered PF_UNIX/PF_LOCAL protocol family
[    0.016392] PCI: CLS 0 bytes, default 64
[    0.018201] workingset: timestamp_bits=46 max_order=17 bucket_order=0
[    0.020114] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.020125] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.045188] Serial: 8250/16550 driver, 3 ports, IRQ sharing disabled
[    0.045792] printk: console [ttyS0] disabled
[    0.065928] 11002000.serial: ttyS0 at MMIO 0x11002000 (irq = 121, base_baud = 2500000) is a ST16650V2
[    0.696370] printk: console [ttyS0] enabled
[    0.701124] mtk_rng 1020f000.trng: registered RNG driver
[    0.701184] hwrng: no data available
[    0.710004] cacheinfo: Unable to detect cache hierarchy for CPU 0
[    0.717915] loop: module loaded
[    0.722192] spi-nand spi0.0: Winbond SPI NAND was found.
[    0.727511] spi-nand spi0.0: 256 MiB, block size: 128 KiB, page size: 2048, OOB size: 128
[    0.735820] 2 fixed-partitions partitions found on MTD device spi0.0
[    0.742157] Creating 2 MTD partitions on "spi0.0":
[    0.746936] 0x000000000000-0x000000400000 : "Bootloader"
[    0.756510] 0x000000400000-0x000010000000 : "UBI_DEV"
[    1.207973] Maxlinear Ethernet GPY211C mdio-bus:06: Firmware Version: 0x886F (release)
[    1.217374] mtk_soc_eth 15100000.ethernet: generated random MAC address 26:cb:9b:3c:93:2a
[    1.226169] mtk_soc_eth 15100000.ethernet eth0: mediatek frame engine at 0xffffffc009300000, irq 129
[    1.235326] mtk_soc_eth 15100000.ethernet: generated random MAC address 12:47:32:04:88:60
[    1.244097] mtk_soc_eth 15100000.ethernet eth1: mediatek frame engine at 0xffffffc009300000, irq 129
[    1.253605] i2c_dev: i2c /dev entries driver
[    1.258781] mtk-wdt 1001c000.watchdog: Watchdog enabled (timeout=31 sec, nowayout=0)
[    1.267097] NET: Registered PF_INET6 protocol family
[    1.272613] Segment Routing with IPv6
[    1.276298] In-situ OAM (IOAM) with IPv6
[    1.280234] NET: Registered PF_PACKET protocol family
[    1.285341] 8021q: 802.1Q VLAN Support v1.8
[    1.290506] pstore: Using crash dump compression: deflate
[    1.334095] mt7530 mdio-bus:1f: no interrupt support
[    1.344602] mt7530 mdio-bus:1f: configuring for fixed/2500base-x link mode
[    1.351907] mt7530 mdio-bus:1f: Link is Up - 2.5Gbps/Full - flow control rx/tx
[    1.361502] mt7530 mdio-bus:1f lan1 (uninitialized): PHY [mt7530-0:01] driver [MediaTek MT7531 PHY] (irq=POLL)
[    1.381303] mt7530 mdio-bus:1f lan2 (uninitialized): PHY [mt7530-0:02] driver [MediaTek MT7531 PHY] (irq=POLL)
[    1.400832] mt7530 mdio-bus:1f lan3 (uninitialized): PHY [mt7530-0:03] driver [MediaTek MT7531 PHY] (irq=POLL)
[    1.420355] mt7530 mdio-bus:1f lan4 (uninitialized): PHY [mt7530-0:04] driver [MediaTek MT7531 PHY] (irq=POLL)
[    1.431147] DSA: tree 0 setup
[    1.434382] ubi0: default fastmap pool size: 100
[    1.439030] ubi0: default fastmap WL pool size: 50
[    1.443803] ubi0: attaching mtd1
[    3.685410] random: crng init done
[    5.568121] ubi0: scanning is finished
[    5.595718] ubi0: attached mtd1 (name "UBI_DEV", size 252 MiB)
[    5.601544] ubi0: PEB size: 131072 bytes (128 KiB), LEB size: 126976 bytes
[    5.608404] ubi0: min./max. I/O unit sizes: 2048/2048, sub-page size 2048
[    5.615170] ubi0: VID header offset: 2048 (aligned 2048), data offset: 4096
[    5.622115] ubi0: good PEBs: 2016, bad PEBs: 0, corrupted PEBs: 0
[    5.628191] ubi0: user volume: 8, internal volumes: 1, max. volumes count: 128
[    5.635390] ubi0: max/mean erase counter: 6/3, WL threshold: 4096, image sequence number: 2037501690
[    5.644498] ubi0: available PEBs: 0, total reserved PEBs: 2016, PEBs reserved for bad PEB handling: 40
[    5.653785] ubi0: background thread "ubi_bgt0d" started, PID 403
[    5.657503] block ubiblock0_6: created from ubi0:6(rootfs)
[    5.665238] ubiblock: device ubiblock0_6 (rootfs) set to be root filesystem
[    5.686989] VFS: Mounted root (squashfs filesystem) readonly on device 254:0.
[    5.694265] Freeing unused kernel memory: 384K
[    5.725430] Run /sbin/init as init process
[    5.729510]   with arguments:
[    5.732461]     /sbin/init
[    5.735153]   with environment:
[    5.738283]     HOME=/
[    5.740628]     TERM=linux
[    6.069450] init: Console is alive
[    6.072925] init: - watchdog -
[    6.762082] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    6.800822] usbcore: registered new interface driver usbfs
[    6.806353] usbcore: registered new interface driver hub
[    6.811674] usbcore: registered new device driver usb
[    6.820411] phy phy-soc:t-phy@11e10000.0: u2 auto load valid efuse: ENABLE with value: 1
[    6.828569] phy phy-soc:t-phy@11e10000.1: u3 auto load valid efuse: ENABLE with value: 1
[    6.836703] phy phy-soc:t-phy@11e10000.2: u2 auto load valid efuse: ENABLE with value: 1
[    6.845433] xhci-mtk 11200000.usb: xHCI Host Controller
[    6.850649] xhci-mtk 11200000.usb: new USB bus registered, assigned bus number 1
[    6.861110] xhci-mtk 11200000.usb: hcc params 0x01403f99 hci version 0x110 quirks 0x0000000000210010
[    6.870257] xhci-mtk 11200000.usb: irq 123, io mem 0x11200000
[    6.876076] xhci-mtk 11200000.usb: xHCI Host Controller
[    6.881286] xhci-mtk 11200000.usb: new USB bus registered, assigned bus number 2
[    6.888667] xhci-mtk 11200000.usb: Host supports USB 3.2 Enhanced SuperSpeed
[    6.896004] hub 1-0:1.0: USB hub found
[    6.899756] hub 1-0:1.0: 2 ports detected
[    6.904005] usb usb2: We don't know the algorithms for LPM for this host, disabling LPM.
[    6.912307] hub 2-0:1.0: USB hub found
[    6.916111] hub 2-0:1.0: 1 port detected
[    6.922526] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    6.936328] init: - preinit -
[    7.432091] mtk_soc_eth 15100000.ethernet eth0: configuring for fixed/2500base-x link mode
[    7.440546] mtk_soc_eth 15100000.ethernet eth0: Link is Up - 2.5Gbps/Full - flow control rx/tx
ip: SIOCSIFHWADDR: Resource busy
[    7.658665] mt7530 mdio-bus:1f lan1: configuring for phy/gmii link mode
Press the [f] key and hit [enter] to enter failsafe mode
Press the [1], [2], [3] or [4] key and hit [enter] to select the debug level
[    9.756991] UBIFS (ubi0:7): default file-system created
[    9.764248] UBIFS (ubi0:7): Mounting in unauthenticated mode
[    9.769993] UBIFS (ubi0:7): background thread "ubifs_bgt0_7" started, PID 547
[    9.969276] UBIFS (ubi0:7): UBIFS: mounted UBI device 0, volume 7, name "rootfs_data"
[    9.977094] UBIFS (ubi0:7): LEB size: 126976 bytes (124 KiB), min./max. I/O unit sizes: 2048 bytes/2048 bytes
[    9.986984] UBIFS (ubi0:7): FS size: 62980096 bytes (60 MiB, 496 LEBs), max 506 LEBs, journal size 3174400 bytes (3 MiB, 25 LEBs)
[    9.998609] UBIFS (ubi0:7): reserved for root: 2974704 bytes (2904 KiB)
[   10.005202] UBIFS (ubi0:7): media format: w5/r0 (latest is w5/r0), UUID E93A0ECA-7E93-44EA-988C-9A50027A8438, small LPT model
[   10.020304] mount_root: overlay filesystem has not been fully initialized yet
[   10.027575] mount_root: switching to ubifs overlay
[   10.037889] overlayfs: null uuid detected in lower fs '/', falling back to xino=off,index=off,nfs_export=off.
[   10.053476] urandom-seed: Seed file not found (/etc/urandom.seed)
[   10.115875] procd: - early -
[   10.118796] procd: - watchdog -
[   10.665650] procd: - watchdog -
[   10.668975] procd: - ubus -
[   10.731798] procd: - init -
Please press Enter to activate this console.
[   10.947306] urngd: v1.0.2 started.
[   10.950707] kmodloader: loading kernel modules from /etc/modules.d/*
[   10.961786] crypto-safexcel 10320000.crypto: EIP97:230(0,1,4,4)-HIA:270(0,5,5),PE:150/433(alg:7fcdfc00)/0/0/0
[   10.976113] Loading modules backported from Linux version v6.1-rc8-0-g76dcd734eca2
[   10.983675] Backport generated by backports.git v5.15.81-1-41-g02e352527db5
[   11.308814] mt7986-wmac 18000000.wmac: HW/SW Version: 0x8a108a10, Build Time: 20221012174648a
[   11.308814]
[   11.631979] mt7986-wmac 18000000.wmac: WM Firmware Version: ____000000, Build Time: 20221012174725
[   11.780497] mt7986-wmac 18000000.wmac: WA Firmware Version: DEV_000000, Build Time: 20221012174937
[   11.927163] mt7986-wmac 18000000.wmac: eeprom load fail, use default bin
[   11.933923] mt7986-wmac 18000000.wmac: Direct firmware load for mediatek/mt7986_eeprom_mt7976_dbdc.bin failed with error -2
[   11.945035] mt7986-wmac 18000000.wmac: Falling back to sysfs fallback for: mediatek/mt7986_eeprom_mt7976_dbdc.bin
[   14.969285] PPP generic driver version 2.4.2
[   14.974166] NET: Registered PF_PPPOX protocol family
[   14.981410] kmodloader: done loading kernel modules from /etc/modules.d/*
[   19.934951] mtk_soc_eth 15100000.ethernet eth0: Link is Down
[   19.946847] mtk_soc_eth 15100000.ethernet eth0: configuring for fixed/2500base-x link mode
[   19.955255] mtk_soc_eth 15100000.ethernet eth0: Link is Up - 2.5Gbps/Full - flow control rx/tx
[   19.955994] device eth0 entered promiscuous mode
[   19.968980] mt7530 mdio-bus:1f lan1: configuring for phy/gmii link mode
[   19.976358] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[   19.984976] br-lan: port 1(lan1) entered blocking state
[   19.990212] br-lan: port 1(lan1) entered disabled state
[   19.996853] device lan1 entered promiscuous mode
[   20.008400] mt7530 mdio-bus:1f lan2: configuring for phy/gmii link mode
[   20.018152] br-lan: port 2(lan2) entered blocking state
[   20.019873] mt7530 mdio-bus:1f lan2: Link is Up - 1Gbps/Full - flow control rx/tx
[   20.023382] br-lan: port 2(lan2) entered disabled state
[   20.037774] device lan2 entered promiscuous mode
[   20.044456] br-lan: port 2(lan2) entered blocking state
[   20.049699] br-lan: port 2(lan2) entered forwarding state
[   20.055343] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
[   20.064125] mt7530 mdio-bus:1f lan3: configuring for phy/gmii link mode
[   20.073542] br-lan: port 3(lan3) entered blocking state
[   20.078857] br-lan: port 3(lan3) entered disabled state
[   20.085926] device lan3 entered promiscuous mode
[   20.094250] mt7530 mdio-bus:1f lan4: configuring for phy/gmii link mode
[   20.103728] br-lan: port 4(lan4) entered blocking state
[   20.108972] br-lan: port 4(lan4) entered disabled state
[   20.116169] device lan4 entered promiscuous mode
[   20.135823] mtk_soc_eth 15100000.ethernet eth1: PHY [mdio-bus:06] driver [Maxlinear Ethernet GPY211C] (irq=POLL)
[   20.145990] mtk_soc_eth 15100000.ethernet eth1: configuring for phy/2500base-x link mode

Ansuel commented:
@bmork ............................. can you by chance create a simple pr to reintroduce them so I can quickly merge them?

dangowrt commented:
@Gingernut1978 I've rebased on top of the fix re-adding the patch for the NAND flash. Please re-test on the AX4200.

mrkiko commented:
Interesting work. Thanks a lot @dangowrt and All

K900 commented:
Run-tested on AX6S.

Gingernut1978 commented:
> @Gingernut1978 I've rebased on top of the fix re-adding the patch for the NAND flash. Please re-test on the AX4200.

```
Starting kernel ...

[    0.000000] Booting Linux on physical CPU 0x0000000000 [0x410fd034]
[    0.000000] Linux version 6.1.34 (asim@ASIM-MS-7D20) (aarch64-openwrt-linux-musl-gcc (OpenWrt GCC 12.3.0 r22878+3-c3e883b308) 12.3.0, GNU ld (GNU Binutils) 2.40.0) #0 SMP Tue Jun 20 07:46:07 2023
[    0.000000] Machine model: ASUS TUF-AX4200
[    0.000000] Zone ranges:
[    0.000000]   DMA      [mem 0x0000000040000000-0x000000005fffffff]
[    0.000000]   DMA32    empty
[    0.000000]   Normal   empty
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000040000000-0x0000000042ffffff]
[    0.000000]   node   0: [mem 0x0000000043000000-0x000000004302ffff]
[    0.000000]   node   0: [mem 0x0000000043030000-0x000000004fbfffff]
[    0.000000]   node   0: [mem 0x000000004fc00000-0x000000004ffbffff]
[    0.000000]   node   0: [mem 0x000000004ffc0000-0x000000005fffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000040000000-0x000000005fffffff]
[    0.000000] psci: probing for conduit method from DT.
[    0.000000] psci: PSCIv1.1 detected in firmware.
[    0.000000] psci: Using standard PSCI v0.2 function IDs
[    0.000000] psci: MIGRATE_INFO_TYPE not supported.
[    0.000000] psci: SMC Calling Convention v1.2
[    0.000000] percpu: Embedded 17 pages/cpu s30312 r8192 d31128 u69632
[    0.000000] pcpu-alloc: s30312 r8192 d31128 u69632 alloc=17*4096
[    0.000000] pcpu-alloc: [0] 0 [0] 1 [0] 2 [0] 3
[    0.000000] Detected VIPT I-cache on CPU0
[    0.000000] CPU features: detected: GIC system register CPU interface
[    0.000000] CPU features: kernel page table isolation disabled by kernel configuration
[    0.000000] alternatives: applying boot alternatives
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 129024
[    0.000000] Kernel command line: rootfstype=squashfs ubi.mtd=UBI_DEV root=/dev/mtdblock6
[    0.000000] Dentry cache hash table entries: 65536 (order: 7, 524288 bytes, linear)
[    0.000000] Inode-cache hash table entries: 32768 (order: 6, 262144 bytes, linear)
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 497380K/524288K available (8320K kernel code, 888K rwdata, 2296K rodata, 448K init, 289K bss, 26908K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=64, Order=0-3, MinObjects=0, CPUs=4, Nodes=1
[    0.000000] rcu: Hierarchical RCU implementation.
[    0.000000]  Tracing variant of Tasks RCU enabled.
[    0.000000] rcu: RCU calculated value of scheduler-enlistment delay is 10 jiffies.
[    0.000000] NR_IRQS: 64, nr_irqs: 64, preallocated irqs: 0
[    0.000000] GICv3: GIC: Using split EOI/Deactivate mode
[    0.000000] GICv3: 640 SPIs implemented
[    0.000000] GICv3: 0 Extended SPIs implemented
[    0.000000] Root IRQ handler: gic_handle_irq
[    0.000000] GICv3: GICv3 features: 16 PPIs
[    0.000000] GICv3: CPU0: found redistributor 0 region 0:0x000000000c080000
[    0.000000] rcu: srcu_init: Setting srcu_struct sizes based on contention.
[    0.000000] arch_timer: cp15 timer(s) running at 13.00MHz (phys).
[    0.000000] clocksource: arch_sys_counter: mask: 0xffffffffffffff max_cycles: 0x2ff89eacb, max_idle_ns: 440795202429 ns
[    0.000000] sched_clock: 56 bits at 13MHz, resolution 76ns, wraps every 4398046511101ns
[    0.000110] Calibrating delay loop (skipped), value calculated using timer frequency.. 26.00 BogoMIPS (lpj=130000)
[    0.000117] pid_max: default: 32768 minimum: 301
[    0.000326] Mount-cache hash table entries: 1024 (order: 1, 8192 bytes, linear)
[    0.000334] Mountpoint-cache hash table entries: 1024 (order: 1, 8192 bytes, linear)
[    0.000989] cacheinfo: Unable to detect cache hierarchy for CPU 0
[    0.001299] cblist_init_generic: Setting adjustable number of callback queues.
[    0.001305] cblist_init_generic: Setting shift to 2 and lim to 1.
[    0.001398] rcu: Hierarchical SRCU implementation.
[    0.001399] rcu:     Max phase no-delay instances is 1000.
[    0.001691] smp: Bringing up secondary CPUs ...
[    0.001954] Detected VIPT I-cache on CPU1
[    0.001995] cacheinfo: Unable to detect cache hierarchy for CPU 1
[    0.002002] GICv3: CPU1: found redistributor 1 region 0:0x000000000c0a0000
[    0.002027] CPU1: Booted secondary processor 0x0000000001 [0x410fd034]
[    0.002293] Detected VIPT I-cache on CPU2
[    0.002321] cacheinfo: Unable to detect cache hierarchy for CPU 2
[    0.002325] GICv3: CPU2: found redistributor 2 region 0:0x000000000c0c0000
[    0.002336] CPU2: Booted secondary processor 0x0000000002 [0x410fd034]
[    0.002590] Detected VIPT I-cache on CPU3
[    0.002618] cacheinfo: Unable to detect cache hierarchy for CPU 3
[    0.002622] GICv3: CPU3: found redistributor 3 region 0:0x000000000c0e0000
[    0.002631] CPU3: Booted secondary processor 0x0000000003 [0x410fd034]
[    0.002655] smp: Brought up 1 node, 4 CPUs
[    0.002659] SMP: Total of 4 processors activated.
[    0.002661] CPU features: detected: 32-bit EL0 Support
[    0.002663] CPU features: detected: CRC32 instructions
[    0.002686] CPU features: emulated: Privileged Access Never (PAN) using TTBR0_EL1 switching
[    0.002689] CPU: All CPU(s) started at EL2
[    0.002690] alternatives: applying system-wide alternatives
[    0.005158] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.005173] futex hash table entries: 1024 (order: 4, 65536 bytes, linear)
[    0.005315] pinctrl core: initialized pinctrl subsystem
[    0.005805] NET: Registered PF_NETLINK/PF_ROUTE protocol family
[    0.006066] DMA: preallocated 128 KiB GFP_KERNEL pool for atomic allocations
[    0.006086] DMA: preallocated 128 KiB GFP_KERNEL|GFP_DMA pool for atomic allocations
[    0.006104] DMA: preallocated 128 KiB GFP_KERNEL|GFP_DMA32 pool for atomic allocations
[    0.006428] thermal_sys: Registered thermal governor 'fair_share'
[    0.006431] thermal_sys: Registered thermal governor 'bang_bang'
[    0.006432] thermal_sys: Registered thermal governor 'step_wise'
[    0.006434] thermal_sys: Registered thermal governor 'user_space'
[    0.006474] ASID allocator initialised with 65536 entries
[    0.006827] pstore: Registered ramoops as persistent store backend
[    0.006830] ramoops: using 0x10000@0x42ff0000, ecc: 0
[    0.014251] cryptd: max_cpu_qlen set to 1000
[    0.015133] SCSI subsystem initialized
[    0.015206] libata version 3.00 loaded.
[    0.016026] clocksource: Switched to clocksource arch_sys_counter
[    0.016576] NET: Registered PF_INET protocol family
[    0.016664] IP idents hash table entries: 8192 (order: 4, 65536 bytes, linear)
[    0.017045] tcp_listen_portaddr_hash hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.017055] Table-perturb hash table entries: 65536 (order: 6, 262144 bytes, linear)
[    0.017061] TCP established hash table entries: 4096 (order: 3, 32768 bytes, linear)
[    0.017082] TCP bind hash table entries: 4096 (order: 5, 131072 bytes, linear)
[    0.017174] TCP: Hash tables configured (established 4096 bind 4096)
[    0.017239] UDP hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.017250] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.017343] NET: Registered PF_UNIX/PF_LOCAL protocol family
[    0.017361] PCI: CLS 0 bytes, default 64
[    0.018232] workingset: timestamp_bits=46 max_order=17 bucket_order=0
[    0.021075] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.021083] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.046554] Serial: 8250/16550 driver, 3 ports, IRQ sharing disabled
[    0.047259] printk: console [ttyS0] disabled
[    0.067437] 11002000.serial: ttyS0 at MMIO 0x11002000 (irq = 116, base_baud = 2500000) is a ST16650V2
[    0.752468] printk: console [ttyS0] enabled
[    0.757258] mtk_rng 1020f000.rng: registered RNG driver
[    0.762699] cacheinfo: Unable to detect cache hierarchy for CPU 0
[    0.771208] loop: module loaded
[    0.775569] spi-nand spi0.0: Winbond SPI NAND was found.
[    0.780890] spi-nand spi0.0: 256 MiB, block size: 128 KiB, page size: 2048, OOB size: 128
[    0.789223] 2 fixed-partitions partitions found on MTD device spi0.0
[    0.795559] Creating 2 MTD partitions on "spi0.0":
[    0.800347] 0x000000000000-0x000000400000 : "Bootloader"
[    0.809923] 0x000000400000-0x000010000000 : "UBI_DEV"
[    1.236102] Maxlinear Ethernet GPY211C mdio-bus:06: Firmware Version: 8.111 (0x886F)
[    1.244420] mtk_soc_eth 15100000.ethernet: generated random MAC address 66:a7:4a:9f:ff:80
[    1.253324] mtk_soc_eth 15100000.ethernet eth0: mediatek frame engine at 0xffffffc009380000, irq 118
[    1.262498] mtk_soc_eth 15100000.ethernet: generated random MAC address 8a:6d:f2:e5:16:76
[    1.271358] mtk_soc_eth 15100000.ethernet eth1: mediatek frame engine at 0xffffffc009380000, irq 118
[    1.280888] i2c_dev: i2c /dev entries driver
[    1.286297] mtk-wdt 1001c000.watchdog: Watchdog enabled (timeout=31 sec, nowayout=0)
[    1.294722] NET: Registered PF_INET6 protocol family
[    1.300411] Segment Routing with IPv6
[    1.304082] In-situ OAM (IOAM) with IPv6
[    1.308050] NET: Registered PF_PACKET protocol family
[    1.313162] 8021q: 802.1Q VLAN Support v1.8
[    1.318624] pstore: Using crash dump compression: deflate
[    1.364056] mt7530-mdio mdio-bus:1f: no interrupt support
[    1.375578] mt7530-mdio mdio-bus:1f: configuring for fixed/2500base-x link mode
[    1.384539] mt7530-mdio mdio-bus:1f: Link is Up - 2.5Gbps/Full - flow control rx/tx
[    1.393767] MediaTek MT7531 PHY mt7530-0:01: Configure LED registers (num=14)
[    1.408296] mt7530-mdio mdio-bus:1f lan1 (uninitialized): PHY [mt7530-0:01] driver [MediaTek MT7531 PHY] (irq=POLL)
[    1.428389] MediaTek MT7531 PHY mt7530-0:02: Configure LED registers (num=14)
[    1.442920] mt7530-mdio mdio-bus:1f lan2 (uninitialized): PHY [mt7530-0:02] driver [MediaTek MT7531 PHY] (irq=POLL)
[    1.462699] MediaTek MT7531 PHY mt7530-0:03: Configure LED registers (num=14)
[    1.477235] mt7530-mdio mdio-bus:1f lan3 (uninitialized): PHY [mt7530-0:03] driver [MediaTek MT7531 PHY] (irq=POLL)
[    1.496976] MediaTek MT7531 PHY mt7530-0:04: Configure LED registers (num=14)
[    1.511505] mt7530-mdio mdio-bus:1f lan4 (uninitialized): PHY [mt7530-0:04] driver [MediaTek MT7531 PHY] (irq=POLL)
[    1.522676] device eth0 entered promiscuous mode
[    1.527314] DSA: tree 0 setup
[    1.530558] ubi0: default fastmap pool size: 100
[    1.535162] ubi0: default fastmap WL pool size: 50
[    1.539967] ubi0: attaching mtd1
[    4.128122] ------------[ cut here ]------------
[    4.132733] WARNING: CPU: 1 PID: 1 at spi_mem_check_op+0xd8/0x110
[    4.138814] Modules linked in:
[    4.141855] CPU: 1 PID: 1 Comm: swapper/0 Not tainted 6.1.34 #0
[    4.147757] Hardware name: ASUS TUF-AX4200 (DT)
[    4.152268] pstate: 80400005 (Nzcv daif +PAN -UAO -TCO -DIT -SSBS BTYPE=--)
[    4.159209] pc : spi_mem_check_op+0xd8/0x110
[    4.163462] lr : spi_mem_check_op+0x7c/0x110
[    4.167715] sp : ffffffc008c5b560
[    4.171013] x29: ffffffc008c5b560 x28: 0000000000000000 x27: 0000000000000000
[    4.178129] x26: 00000000ffffffb6 x25: ffffff8000c3be80 x24: ffffff8000073800
[    4.185244] x23: 0000000000000000 x22: ffffff80003ab000 x21: ffffff8000bc8980
[    4.192360] x20: 0000000000000800 x19: ffffffc008c5b898 x18: 0000000000000014
[    4.199475] x17: 0000000040f6d175 x16: 0000000026a71c1a x15: 00000000cde4ee71
[    4.206590] x14: 000accf1a2f5fe72 x13: c4874af300000000 x12: 0000000000000000
[    4.213705] x11: 2c04000000000000 x10: 0000000000000000 x9 : 0000000000000000
[    4.220820] x8 : ffffffc008c5b870 x7 : 0000000000000000 x6 : 000000000000003f
[    4.227935] x5 : 0000000000000040 x4 : ffffffffffffffd0 x3 : 0000000000000030
[    4.235050] x2 : ffffffc008c5b897 x1 : ffffffc008c5c000 x0 : 0000000000000000
[    4.242167] Call trace:
[    4.244598]  spi_mem_check_op+0xd8/0x110
[    4.248505]  spi_mem_exec_op+0x3c/0x3f0
[    4.252325]  w25n02kv_ecc_get_status+0x78/0xd0
[    4.256752]  spinand_ondie_ecc_finish_io_req+0x40/0xd4
[    4.261875]  nand_ecc_finish_io_req+0x20/0x30
[    4.266219]  spinand_read_page+0x114/0x1e0
[    4.270297]  spinand_mtd_read+0x1b0/0x2fc
[    4.274289]  mtd_read_oob_std+0x58/0x80
[    4.278112]  mtd_read_oob+0x8c/0x150
[    4.281674]  mtd_read+0x50/0x70
[    4.284800]  ubi_io_read+0xb4/0x310
[    4.288275]  ubi_io_read_vid_hdr+0x50/0x1ec
[    4.292441]  scan_peb+0x138/0x720
[    4.295742]  scan_all+0x64/0xacc
[    4.298955]  ubi_attach+0x290/0x340
[    4.302428]  ubi_attach_mtd_dev+0x52c/0xbd4
[    4.306594]  ubi_init+0x138/0x320
[    4.309898]  do_one_initcall+0x4c/0x1d0
[    4.313719]  kernel_init_freeable+0x230/0x294
[    4.318062]  kernel_init+0x20/0x120
[    4.321538]  ret_from_fork+0x10/0x20
[    4.325098] ---[ end trace 0000000000000000 ]---
[    4.329707] ubi0: fixable bit-flip detected at PEB 1260
[    4.436021] random: crng init done
[    5.885379] ubi0: scanning is finished
[    5.913201] ubi0: attached mtd1 (name "UBI_DEV", size 252 MiB)
[    5.919043] ubi0: PEB size: 131072 bytes (128 KiB), LEB size: 126976 bytes
[    5.925896] ubi0: min./max. I/O unit sizes: 2048/2048, sub-page size 2048
[    5.932665] ubi0: VID header offset: 2048 (aligned 2048), data offset: 4096
[    5.939607] ubi0: good PEBs: 2016, bad PEBs: 0, corrupted PEBs: 0
[    5.945680] ubi0: user volume: 8, internal volumes: 1, max. volumes count: 128
[    5.952884] ubi0: max/mean erase counter: 13/8, WL threshold: 4096, image sequence number: 2037501690
[    5.962080] ubi0: available PEBs: 0, total reserved PEBs: 2016, PEBs reserved for bad PEB handling: 40
[    5.971366] ubi0: background thread "ubi_bgt0d" started, PID 415
[    5.978787] block ubiblock0_6: created from ubi0:6(rootfs)
[    5.984261] ubiblock: device ubiblock0_6 (rootfs) set to be root filesystem
[    5.994121] /dev/root: Can't open blockdev
[    5.998246] VFS: Cannot open root device "mtdblock6" or unknown-block(0,0): error -6
[    6.005965] Please append a correct "root=" boot option; here are the available partitions:
[    6.014303] 1f00            4096 mtdblock0
[    6.014308]  (driver?)
[    6.020820] 1f01          258048 mtdblock1
[    6.020823]  (driver?)
[    6.027338] fe00           18228 ubiblock0_6
[    6.027340]  (driver?)
[    6.034022] Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(0,0)
[    6.042261] SMP: stopping secondary CPUs
[    6.046168] Kernel Offset: disabled
[    6.049639] CPU features: 0x00000,00000004,0000400b
[    6.054498] Memory Limit: none
[    6.059617] Rebooting in 1 seconds..

Still bootlooping. Seems to have gotten further after latest fixes.

bmork commented:
I guess this is because  target/linux/mediatek/patches-5.15/901-arm-add-cmdline-override.patch was dropped.  But I don't understand why that is.  @dangowrt  - was this intentional?  If not,  is this the only one missing, or do we need to go through all the patches?

dangowrt commented:
@bmork The logs clearly indicate that CMDLINE_OVERRIDE doesn't work, but the option [is enabled](https://github.com/openwrt/openwrt/pull/12937/commits/9d18eb9052c5054387dfcd491b2e441d96408627#diff-a2d5a98464639b71a7c68b8f83bf785b8e0cb750f9d3a8e4218b96ea6cdcbba2R57) and the patch was not dropped but [is present](https://github.com/openwrt/openwrt/pull/12937/commits/9d18eb9052c5054387dfcd491b2e441d96408627#diff-87c7fb5681b17ffbb191e99e0676ee0fe1da3ebeb4c90db42c1618a790d6d91f).
The original patch for Linux 5.15 did not apply cleanly, so I copied it from another target already updated to Linux 6.1 which also had it.

I can't really look into it now, but guess something went wrong along those lines.

bmork commented:
> The logs clearly indicate that CMDLINE_OVERRIDE doesn't work, but the option [is enabled](https://github.com/openwrt/openwrt/pull/12937/commits/9d18eb9052c5054387dfcd491b2e441d96408627#diff-a2d5a98464639b71a7c68b8f83bf785b8e0cb750f9d3a8e4218b96ea6cdcbba2R57) and the patch was not dropped but [is present](https://github.com/openwrt/openwrt/pull/12937/commits/9d18eb9052c5054387dfcd491b2e441d96408627#diff-87c7fb5681b17ffbb191e99e0676ee0fe1da3ebeb4c90db42c1618a790d6d91f).

Ooops, wrong branch.  Apologies.  You're quite right, of course.

bmork commented:
Yes, the patch was there but it is missing the last chunk defining the symbol for arm64.  So it's dropped from the kernel .config when building for arm64.

You might want to fold this into your 6.1 commit: https://github.com/bmork/openwrt/commit/d1733697fffc308a86fc8d126fc62e0f8959e53d

Gingernut1978 commented:
@dangowrt @bmork Asus tuf4200ax booted and running great after latest force push fixes.

Thank you all for your help

agners commented:
Tested on Netgear WAX220 as well, looks good. For reference, dmesg of kernel: [mediatek-6.1.34-on-netgear-wax220.txt](https://github.com/openwrt/openwrt/files/11861099/mediatek-6.1.34-on-netgear-wax220.txt)

PolynomialDivision commented:
Would be nice if you could divide the testing kernel commit into two (or even three) commits. In the first commit you just copy config and patches from 6.1 and in the other you do the necessary changes. That makes it easier to review and see the diff.

For example (bmips did 3):
- https://github.com/openwrt/openwrt/commit/a937eef305ca3ce0bb945185f3c2914478f14c4a
- https://github.com/openwrt/openwrt/commit/959d3f0582d800bf803cee3edffc82bf006a7543
- https://github.com/openwrt/openwrt/commit/f6c02b014d9c57f25fb8a3a0dfcf52894492bd15

danpawlik commented:
After 2 hours, works well on Xiaomi AX3200.

dangowrt committed:
mediatek: copy patches and files for Linux 6.1

First step only copies patches-5.15 and files-5.15 to
patches-6.1 and files-6.1 respectively.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
mediatek: adapt files and patches for Linux 6.1

With Linux 6.1 many of our downstream patches and out-of-tree files
can be removed or at least replaced by backported upstream commits.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>
[fix CMDLINE_OVERRIDE for arm64]
Signed-off-by: Bjørn Mork <bjorn@mork.no>

dangowrt committed:
mediatek: copy config-5.15 to config-6.1 for all subtargets

To ease review, first copy all subtargets' kernel config-5.15 to the
to-be-adapted config-6.1.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
mediatek: adapt kernel configuration for Linux 6.1

Update kernel configuration to build Linux 6.1 for all subtargets.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

danpawlik commented:
> My qihoo 360t7 keeps crashing after enabling WED, but everything else works fine. [360t7_wed.txt](https://github.com/openwrt/openwrt/files/11878187/360t7_wed.txt)

Maybe @nbd168 can help as he was an author of WED.

Anteus commented:
@danpawlik Did you set CONFIG_CMDLINE_OVERRIDE while building for Xiaomi AX3200? Want to try this pull request for rt3200 but didn't get it to boot.

patrykk commented:
> @dangowrt @bmork Asus tuf4200ax booted and running great after latest force push fixes.
> 
> Thank you all for your help
Which branch and repo have you used for final tests and could you prepare separate dts file with vendor partitions table 
 like it is with mt7986a-xiaomi-redmi-router-ax6000-stock.dts ? 

best regards,

K900 commented:
FWIW this currently doesn't build non-interactively at least on ax6s, because the kernel config prompts for `CONFIG_CMDLINE_OVERRIDE`.

dangowrt commented:
@K900 Sounds like I either made a mistake when merging @bmork's fix or your checkout is outdated and doesn't yet contain https://github.com/openwrt/openwrt/pull/12937/commits/56ba61b0fbf1dd3ee16961d596ea8d89f9deb0ce

K900 commented:
I am using the latest version of the branch.

Gingernut1978 commented:
T´´´# CONFIG_CMDLINE_OVERRIDE is not set ``` is missing from config-6.1 in the mt7622 target

dangowrt commented:
> It was removed in [8c2b3f5](https://github.com/openwrt/openwrt/pull/12937/commits/8c2b3f56b535e298e796aeeb20484fabffa760ad)

Fixed in https://github.com/openwrt/openwrt/pull/12937/commits/13e9203d18fbb474b2eb9ee7aefe948337ae3f25
@Gingernut1978  Thanks for pointing it out

K900 commented:
OK so I have no idea what is happening here, but I'm pretty sure my network is _somehow_ less stable on 6.1 than on 5.15 on ax6s. I can't find any hard evidence for it, but I get a lot more stuttering when streaming video over the network somehow. Has anyone else noticed anything like this? Is my router gaslighting me?

dangowrt committed:
generic: 6.1: fix uImage.FIT partitions on mtdblock and ubiblock

When refreshing the hack patches for Linux 6.1 the part of the uImage.FIT
partition parser patch which takes care of allowing mtdblock and ubiblock
devices to have partitions has been dropped, supposedly by accident.
Re-add a that part to the patch, so devices using a uImage.FIT filesystem
sub-image as rootfs can work with Linux 6.1.

Fixes: 19a246bb65 ("generic: 6.1: manually refresh hack patches")
Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt commented:
Pushed to main branch: https://github.com/openwrt/openwrt/commit/1321007e1f68c07280b3ca5ba11941cbad7a5312
